add inexpensive guess for sysPrefix when withSysPrefix: false

minrk opened the pull request:
effectively does `which jupyter` without instantiating any subprocesses, and resolves `realpath(jupyter)/../..` for the prefix.

I also removed the `sort` calls from the tests, which hid a bug where the ordering of system data dirs was not correct

Tests pass on OS X with conda Python, as well as virtualenvs.

Another option is to check on `python`, rather than `jupyter`. The `realpath` in that case will give the right answer for homebrew Python, which this doesn't (no change from master). The disadvantage is that it will include Python's prefix whether the Jupyter Python package is installed or not, and there isn't a way to identify _which_ Python without reading the shebang line of the Jupyter script, and I wanted to make sure that no files were opened or processes executed.

If Jupyter is installed with `--user`, this will add the user prefix, which is not on the Python implementation's path, so will generally turn up no results.

minrk committed:
add inexpensive guess for sysPrefix when withSysPrefix: false

effectively does `which jupyter` without instantiating any subprocesses, and resolves `realpath(jupyter)/../..`

I also removed the `sort` calls from the tests, which hid a bug where the ordering of system data dirs was not correct

Tests pass on OS X with conda Python

minrk committed:
case-insensitive comparisons in tests

seems necessary on Windows

rgbkrk commented on index.js:
```diff
@@ -4,9 +4,64 @@
 const path = require('path');
 const home = require('home-dir');
-const sysPrefixPromise = require('sys-prefix-promise')
+const sysPrefixPromise = require('sys-prefix-promise');
```
Thanks for cleaning up my loosey-goosey semi colons while you were at it. :wink:

rgbkrk commented on index.js:
```diff
@@ -4,9 +4,64 @@
+  // based on shutil.which from Python 3.5
+
+  // only run once:
+  if (sysPrefixGuess !== undefined) return sysPrefixGuess;
```
This returns null for all the cases that it couldn't resolve one nicely?

rgbkrk commented on index.js:
```diff
@@ -28,16 +83,25 @@ function configDirs(opts) {
-  return paths.concat(systemConfigDirs());
+  // inexpensive guess, based on location of `jupyter` executable
+  var sysPrefix = guessSysPrefix();
+  var sysPathed = path.join(sysPrefix, 'etc', 'jupyter');
```
👍

rgbkrk commented:
<img width="534" alt="screen shot 2016-08-02 at 3 52 16 pm" src="https://cloud.githubusercontent.com/assets/836375/17348137/25f400a4-58c9-11e6-9664-967140c7a9b8.png">

Is this the case where it works for virtual envs yet not `--user`?

minrk committed:
check for python instead of jupyter

+ not sensitive to `--user` installs
+ more likely to follow symlinks to sys.prefix (homebrew)
- might catch the wrong Python

It should be hard to catch the wrong Python in a way that matters, though.
The only real case is Jupyter installed with Python 3,
no envs (not conda, not virtualenv, not python brew, etc.),
and `which python` is python2
and python3 has an unusual prefix (not /usr/local, not /usr).

It could also happen where a virtualenv is active,
but jupyter is not in the env, not on /usr:/usr/local,
yet still on PATH.

minrk commented:
> Is this the case where it works for virtual envs yet not --user?

Yes, precisely. Because of things like this, I'm not 100% sure it's a good idea to do it this way, however with the increasing use of things like conda to install kernels, I think specs are increasingly going to show up on sys.prefix, so I thought I would take a run at it.

I switched it to check for `python` instead of `jupyter`. This should get the right answer, as long as `which python` returns the Python for which Jupyter is installed. In fact, it should do just as good a job as `sys-prefix-promise`, which also relies on `which python` returning the right thing, so maybe that option should be dropped.

rgbkrk commented:
Would you want to tackle this one again @minrk or should we close this?

minrk committed:
fix guess on Windows

prefix is only one level up instead of two

minrk commented:
Sorry, totally forgot this one. Should be fixed.

rgbkrk commented:
CI not passing, I'll dig on this one later.

lgeiger commented:
The CI failure should be fixed by #19

minrk commented:
rebased after #19, all green.

rgbkrk commented:
Ah, sweet
